### PR TITLE
[clang][dataflow] Allow `DataflowAnalysisContext` to use a non-owned `Solver`.

### PR DESCRIPTION
**Pull request by martinboehme**
For some callers (see change in DataflowAnalysis.h), this is more convenient.

### CODE REVIEW

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h [hunk 1]
```diff
@@ -209,6 +221,9 @@ class DataflowAnalysisContext {
     using DenseMapInfo::isEqual;
   };
 
+  DataflowAnalysisContext(Solver &S, std::unique_ptr<Solver> OwnedSolver,
```
Comment (ymand) [inline clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h:224]: Please add comments either here or on the fields relating S and OwnedSolver.

**Review by ymand — APPROVED** [review]

File: clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h (hunk 1 above)
Comment (martinboehme) [reply clang/include/clang/Analysis/FlowSensitive/DataflowAnalysisContext.h]: Done.

**Review by ymand — APPROVED** [review]

**Comment by martinboehme** [comment]
@ymand PTAL -- I noticed that my initial version had a use-after-move on this line:

```cxx
  DataflowAnalysisContext(std::unique_ptr<Solver> S,
                          Options Opts = Options{
                              /*ContextSensitiveOpts=*/std::nullopt,
                              /*Logger=*/nullptr})
      : DataflowAnalysisContext(*S, std::move(S), Opts) {}
```

This was because the private constructor to which this constructor delegates originally took the `std::unique_ptr<Solver>` by value, and the order in which the arguments are evaluated is indeterminate.

I've now changed the parameter of the private constructor to be an rvalue reference (`std::unique_ptr<Solver> &&`), which resolves this issue as the move only happens inside the constructor.

**Review by Xazax-hun — APPROVED** [review]

**Comment by martinboehme** [comment]
CI failure looks unrelated.

**Comment by joker-eph** [comment]
CI failure wasn't unrelated actually: the CI is consistently broken after this patch (you have got an email about it?), I'll revert for now.

**Comment by joker-eph** [comment]
Actually your premerge failure here was an infra issue, but the CI is still broken after this PR somehow: https://lab.llvm.org/buildbot/#/builders/271/builds/7420

**Comment by martinboehme** [comment]
> Actually your premerge failure here was an infra issue, but the CI is still broken after this PR somehow: https://lab.llvm.org/buildbot/#/builders/271/builds/7420
> 
> ```
> # .---command stderr------------
> # | error: 'expected-error' diagnostics seen but not expected: 
> # |   (frontend): '-fsyntax-only' action ignored; '-ast-print' action specified previously
> # `-----------------------------
> ```

This is a failure in the test clang\test\OpenMP\target_ast_print.cpp, correct?

Are you sure this PR is the culprit? It touches only the dataflow framework, which is not used in OpenMP or the Clang compiler itself.

**Pull request merged**